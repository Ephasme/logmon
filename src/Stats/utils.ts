import { IBasicState, StateBySections, IBatchState } from "./types";
import { ILogLine } from "../Models/ILogLine";
import { Set, Map } from "immutable";

export type Adder<T> = (state1: T, state2: T) => T;
export type BasicStateAdder = Adder<IBasicState>;
export type StateBySectionsAdder = Adder<StateBySections>;
export type BatchStateAdder = Adder<IBatchState>;

export const add: BasicStateAdder = (state1, state2) => ({
    errors: state2.errors + state1.errors,
    hits: state2.hits + state1.hits,
    traffic: state2.traffic + state1.traffic,
});

export const mergeSections: StateBySectionsAdder = (sections1, sections2) =>
    sections1.mergeWith((prev, next) => add(prev, next), sections2);

export const mergeBatches: BatchStateAdder = (batch1, batch2) => ({
    sections: mergeSections(batch1.sections, batch2.sections),
    ...add(batch1, batch2),
});

export function groupBySections(logs: ILogLine[]): Map<string, Set<ILogLine>> {
    return logs
        .filter((x) => x.request.routeSegments.length > 0)
        .map((x) => ({
            id: x.request.routeSegments[0],
            ...x,
        }))
        .reduce((prev, cur) => {
            const newSet = prev.get(cur.id, Set<ILogLine>()).add(cur);
            return prev.set(cur.id, newSet);
        }, Map<string, Set<ILogLine>>());
}