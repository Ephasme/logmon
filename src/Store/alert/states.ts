import { List } from "immutable";

import { ILogLine } from "../../LogWatcher";

export const OVERLOADING = "OVERLOADING";
export type OverloadingStatus = {
    type: typeof OVERLOADING,
    since: Date,
    hits: number,
};

export const RECOVERING = "RECOVERING"
export type RecoveringStatus = {
    type: typeof RECOVERING,
    since: Date,
};

export const IDLE = "IDLE"
export type IdleStatus = {
    type: typeof IDLE,
    since: Date,
};

export type AnyStatus =
    | IdleStatus
    | RecoveringStatus
    | OverloadingStatus; 

export type AlertState = Readonly<{
    status: AnyStatus,
    logs: List<ILogLine>,
}>;

export const defaultAlertStateFactory: () => AlertState = () => ({
    status: { type: IDLE, since: new Date() },
    logs: List(),
});
