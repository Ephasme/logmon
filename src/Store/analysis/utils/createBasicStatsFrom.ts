import { List } from "immutable";
import { ILogLine } from "../../../LogWatcher";
import { computeTimeGap } from "../../common/computeTimeGap";
import { Seconds, Sec } from "../../../Utils/units";

export interface IBasicStats {
    readonly hits: number,
    readonly traffic: number,
    readonly errors: number,
    readonly timespan: Seconds,
}

export const defaultBasicStatsFactory: () => IBasicStats = () => ({
    hits: 0,
    traffic: 0,
    errors: 0,
    timespan: Sec(0),
});

export const createBasicStatsFrom: (logs: List<ILogLine>) => IBasicStats = (logs) => {
    const stats = {
        traffic: 0,
        errors: 0,
    }
    for (const log of logs) {
        stats.errors += log.result >= 200 && log.result <= 299 ? 0 : 1;
        stats.traffic += log.packet;
    }
    return {
        ...stats,
        hits: logs.size,
        timespan: computeTimeGap(logs, Sec(Infinity)),
    };
};