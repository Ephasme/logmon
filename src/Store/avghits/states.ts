import { List } from "immutable";
import { ILogLine } from "../../LogWatcher";

export interface IAlertMessage {
    type: "alert";
    hits: number;
    time: Date;
}

export interface IInfoMessage {
    type: "info";
    time: Date;
}

export type AnyMessage = IAlertMessage | IInfoMessage;

export type AvgHitsState = Readonly<{
    logs: List<ILogLine>;
    avgHitsPerSeconds: number;
    status: "triggered" | "idle";
    messages: List<AnyMessage>;
}>;

export const defaultAvgHitsStateFactory: () => AvgHitsState = () => ({
    logs: List(),
    status: "idle",
    avgHitsPerSeconds: 0,
    messages: List(),
});
