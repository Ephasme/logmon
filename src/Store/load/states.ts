import { List } from "immutable";

import { ILogLine } from "../../LogWatcher";
import { ISeconds, Sec } from "../../Utils/units";

export interface IAlertMessage {
    type: "alert";
    hits: number;
    time: Date;
}

export interface IInfoMessage {
    type: "info";
    time: Date;
}

export type AnyMessage = IAlertMessage | IInfoMessage;

export type LoadState = Readonly<{
    overloadDuration: ISeconds,
    status: "TRIGGERED" | "IDLE",
    messages: List<AnyMessage>,
    logs: List<ILogLine>,
}>;

export const defaultLoadStateFactory: () => LoadState = () => ({
    overloadDuration: Sec(0),
    status: "IDLE",
    logs: List(),
    messages: List(),
});
