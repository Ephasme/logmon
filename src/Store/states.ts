import { ILogLine } from "../LogWatcher";
import { List } from "immutable";

export interface IAlertMessage {
    type: "alert";
    hits: number;
    time: Date;
}

export interface IRecoverMessage {
    type: "recover";
    time: Date;
}

export type AnyMessage = IAlertMessage | IRecoverMessage;

export type DataState = Readonly<{
    overloadingStatus: "IDLE" | "TRIGGERED",
    message: AnyMessage | null,
    currentHitsPerSeconds: number,
    timespan: number;
}>;

export type RootState = Readonly<{
    logs: List<ILogLine>;
    data: DataState;
}>;

export const defaultStateFactory: () => RootState = () => ({
    logs: List<ILogLine>(),
    data: {
        overloadingStatus: "IDLE",
        message: null,
        currentHitsPerSeconds: 0,
        timespan: 0,
    },
});
