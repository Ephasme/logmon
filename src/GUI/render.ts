import { IState, IBatchState, IBasicState } from "../Stats/types";
import { Map } from "immutable";
import * as blessed from "blessed";

export const maybeGetBatch = (batch: IBatchState): IBatchState | null =>
    batch.sections.size > 0 ? batch : null;

export function getBatch(state: IState) {
    return maybeGetBatch(state.currentBatch) ||
           maybeGetBatch(state.lastValidBatch);
}

export function batchToTableData(state?: [string, IBasicState][]): string[][] {
    return (state || []).reduce((acc, [idx, el]) => {
        return [ ...acc, [ idx, el.hits.toString(), el.traffic.toString(), el.errors.toString() ]];
    }, [["section", "hits", "traffic", "errors"]]);
}

export interface IGui {
    render(): void;
}

export function createGui() {
    const screen = blessed.screen({
        smartCSR: true,
        title: "LogMon",
    });

    const alert = blessed.box({
        parent: screen,
        top: 0,
        style: {
            bg: "red",
            fg: "white",
        },
        content: "ALERT:\nSome alert",
    })

    const header = blessed.box({
        parent: screen,
        top: 2,
        content: "Welcome to LogMon - an access log monitoring console application.\n" +
                 "\n" +
                 "You can use options to customize the timing, \n" +
                 "and the alerting behaviour (see --help command for more infos).\n" +
                 "\n" +
                 "> Made with love by Ephasme... <3\n" +
                 "> Press ESC, a, or C-c to quit.",

    });

    const globalStats = blessed.box({
        top: <number>header.top + 7,
        parent: screen,
    });

    const listtableTitle = blessed.box({
        parent: screen,
        top: <number>globalStats.top + 8,
        content: "Last batch statistics:"
    });
    
    const listtableDate = blessed.box({
        parent: screen,
        top: <number>listtableTitle.top + 1,
        content: "Last update:",
    });

    const listtable = blessed.listtable({
        parent: screen,
        border: "line",
        top: <number>listtableDate.top + 1,
        scrollable: true,
        height: "100%-22",
        width: "50%",
        data: [],
    });

    screen.key(['escape', 'q', 'C-c'], () => process.exit(0));

    type BasicStateWithKey = IBasicState & { key: string };
    
    function most<T>(all: Map<string, BasicStateWithKey>, selector: (batch: BasicStateWithKey) => T, name: string) {
        const mostVisited = all.maxBy(selector);
        if (mostVisited) {
            return `${mostVisited.key} (${selector(mostVisited)} ${name})`
        }
        return "";
    }

    const alertMessage = (value: number, now: Date) =>
        `High traffic generated an alert - hits = ${value}, triggered at ${now}`;

    const recoverMessage = (now: Date) =>
        `Recovered from high traffic at ${now}`;

    return {
        render: (state: IState) => {
            const batch = getBatch(state);

            if (state.alert.status === "off") {
                const message = state.alert.message.get(0);
                if (message && message.type === "alert") {
                    alert.setContent(`Alert:\n\t${alertMessage(message.hits, message.time)}`);
                    alert.show();
                } else if (message && message.type === "recover") {
                    alert.setContent(`Alert:\n\t${recoverMessage(message.time)}`);
                    alert.show();
                }
            } else {
                alert.hide();
            }

            if (batch) {
                listtable.setData(batchToTableData(batch.sections.toArray()));
            } else {
                listtable.setData(batchToTableData());
            }
            listtableDate.setContent(state.lastUpdated.toLocaleString());

            const allBatchesWithKeys = state.allBatches.sections
                .map((x, key) => ({ ...x, key }));

            const mostVisits = most(allBatchesWithKeys, (x) => x.hits, "hit(s)");
            const mostErrorProne = most(allBatchesWithKeys, (x) => x.errors, "error(s)");
            globalStats.setContent(`
Global stats:
    
    Most visited section: ${mostVisits}
    Most buggy section: ${mostErrorProne}
    Total hits: ${state.allBatches.hits} hit(s)
    Total traffic: ${state.allBatches.traffic} b`)

            screen.render();
        }
    }
}

const gui = createGui();

export function render(state: IState) {
    gui.render(state);
}