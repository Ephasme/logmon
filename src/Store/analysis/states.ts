import { List, Map } from "immutable";
import { ILogLine } from "../../LogWatcher";
import { defaultBasicStatsFactory, IBasicStats } from "./utils/createBasicStatsFrom";

export interface IAnalysisState {
    sections: Map<string, IBasicStats>;
    totalAll: IBasicStats;
    totalBatch: IBasicStats;
    currentBatch: List<ILogLine>;
}

export const defaultAnalysisStateFactory: () => IAnalysisState = () => ({
    currentBatch: List(),
    totalAll: defaultBasicStatsFactory(),
    totalBatch: defaultBasicStatsFactory(),
    sections: Map(),
});
