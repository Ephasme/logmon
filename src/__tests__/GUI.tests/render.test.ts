import { List, Map } from "immutable";
import { createGui } from "../../GUI/render";
import { IBasicStats } from "../../Store/analysis/utils/createBasicStatsFrom";
import { AnyMessage, AvgHitsState, defaultAvgHitsStateFactory } from "../../Store/avghits/states";
import { defaultStateFactory, RootState } from "../../Store/states";
import { Sec } from "../../Utils/units";

export const createState = (seed: Partial<AvgHitsState>): RootState => {
    const someState: RootState = defaultStateFactory();
    return {
        ...someState,
        avgHits: {
            ...defaultAvgHitsStateFactory(),
            ...seed,
        },
    };
};
export const createSettings = () => ({
    filename: "filename",
    maxHitsPerSeconds: 12,
    maxOverloadDuration: 5,
    secondsPerRefresh: 2,
});

it("should display many messages", () => {
    const result: string[] = [];
    const gui = createGui(jest.fn(), (input) => { result.push(input); });
    gui.render(createState({
        status: "triggered",
        messages: List<AnyMessage>([
            {type: "alert", hits: 15, time: new Date(2015, 1, 2, 4, 1, 3)},
            {type: "info", time: new Date(2015, 1, 2, 5, 1, 3)},
            {type: "alert", hits: 15, time: new Date(2015, 1, 2, 6, 1, 3)},
        ]),
    }), new Date(2015, 1, 1, 1, 12, 51), "filename", 10, Sec(5));
    expect(result.join("\n")).toMatchSnapshot();
});

it("should display an alert message when message is alert", () => {
    const result: string[] = [];
    const gui = createGui(jest.fn(), (input) => { result.push(input); });
    gui.render(createState({
        status: "triggered",
        messages: List([{type: "alert", hits: 15, time: new Date(2015, 1, 2, 4, 1, 3)}]),
    }), new Date(2015, 1, 1, 1, 12, 51), "file", 10, Sec(5));
    expect(result.join("\n")).toMatchSnapshot();
});

it("should display recovering message when message is recovering", () => {
    const result: string[] = [];
    const gui = createGui(jest.fn(), (input) => { result.push(input); });
    gui.render(createState({
        status: "triggered",
        messages: List([{type: "info", time: new Date(2015, 1, 2, 4, 1, 3)}]),
    }), new Date(2015, 1, 1, 1, 12, 51), "file", 10, Sec(5)),
    expect(result.join("\n")).toMatchSnapshot();
});

it("should display sections when filled", () => {
    const result: string[] = [];
    const state: RootState = {
        avgHits: defaultAvgHitsStateFactory(),
        analysis: {
            currentBatch: List([]),
            sections: Map<string, IBasicStats>([
                ["test1", {
                    errors: 12,
                    hits: 8,
                    timespan: Sec(73),
                    traffic: 91,
                }],
                ["test2", {
                    errors: 57,
                    hits: 14,
                    timespan: Sec(74),
                    traffic: 123,
                }],
            ]),
            totalAll: {
                errors: 50,
                hits: 55,
                timespan: Sec(127),
                traffic: 66,
            },
            totalBatch: {
                errors: 76,
                hits: 22,
                timespan: Sec(33),
                traffic: 44,
            },
        },
    };
    const gui = createGui(jest.fn(), (input) => { result.push(input); });
    gui.render(state, new Date(2015, 1, 1, 1, 12, 51), "file", 10, Sec(5));
    expect(result.join("\n")).toMatchSnapshot();
});
