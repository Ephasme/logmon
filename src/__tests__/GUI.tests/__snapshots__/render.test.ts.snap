// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display an alert message when message is alert 1`] = `
"/!\\\\ Alert: High traffic generated an alert - hits = 15, triggered at Mon, 02 Feb 2015 03:01:03 GMT

Welcome to LogMon - an access log monitoring console application.

You can use options to customize the timing, 
and the alerting behaviour (see --help command for more infos).

Last updated at Sun, 01 Feb 2015 00:12:51 GMT


Settings:
    - Overload duration:     0/12
    - Max hits per second:   10
    - Monitored file:        file

Global stats:
    - Most visited section:  -
    - Most buggy section:    -
    - Total hits:            0 hit(s)
    - Avg hits/s:            0.00
    - Total traffic:         0 b

Current batch:
    - Current batch traffic: 0
    - Current batch hits/s:  0.00

Sections details:


> Handcrafted with care for Datadog by Ephasme... <3
> Press C-c to quit."
`;

exports[`should display many messages 1`] = `
"/!\\\\ Alert: High traffic generated an alert - hits = 15, triggered at Mon, 02 Feb 2015 03:01:03 GMT
[o] Info: Recovered from high traffic at Mon, 02 Feb 2015 04:01:03 GMT
/!\\\\ Alert: High traffic generated an alert - hits = 15, triggered at Mon, 02 Feb 2015 05:01:03 GMT

Welcome to LogMon - an access log monitoring console application.

You can use options to customize the timing, 
and the alerting behaviour (see --help command for more infos).

Last updated at Sun, 01 Feb 2015 00:12:51 GMT


Settings:
    - Overload duration:     0/12
    - Max hits per second:   10
    - Monitored file:        file

Global stats:
    - Most visited section:  -
    - Most buggy section:    -
    - Total hits:            0 hit(s)
    - Avg hits/s:            0.00
    - Total traffic:         0 b

Current batch:
    - Current batch traffic: 0
    - Current batch hits/s:  0.00

Sections details:


> Handcrafted with care for Datadog by Ephasme... <3
> Press C-c to quit."
`;

exports[`should display recovering message when message is recovering 1`] = `
"[o] Info: Recovered from high traffic at Mon, 02 Feb 2015 03:01:03 GMT

Welcome to LogMon - an access log monitoring console application.

You can use options to customize the timing, 
and the alerting behaviour (see --help command for more infos).

Last updated at Sun, 01 Feb 2015 00:12:51 GMT


Settings:
    - Overload duration:     0/12
    - Max hits per second:   10
    - Monitored file:        file

Global stats:
    - Most visited section:  -
    - Most buggy section:    -
    - Total hits:            0 hit(s)
    - Avg hits/s:            0.00
    - Total traffic:         0 b

Current batch:
    - Current batch traffic: 0
    - Current batch hits/s:  0.00

Sections details:


> Handcrafted with care for Datadog by Ephasme... <3
> Press C-c to quit."
`;

exports[`should display sections when filled 1`] = `
"Welcome to LogMon - an access log monitoring console application.

You can use options to customize the timing, 
and the alerting behaviour (see --help command for more infos).

Last updated at Sun, 01 Feb 2015 00:12:51 GMT


Settings:
    - Overload duration:     0/12
    - Max hits per second:   10
    - Monitored file:        file

Global stats:
    - Most visited section:  test2 (14 hit(s))
    - Most buggy section:    test2 (57 error(s))
    - Total hits:            55 hit(s)
    - Avg hits/s:            0.43
    - Total traffic:         66 b

Current batch:
    - Current batch traffic: 44
    - Current batch hits/s:  0.67

Sections details:

test1: 91 bytes in 8 hit(s) (12 error(s))
test2: 123 bytes in 14 hit(s) (57 error(s))

> Handcrafted with care for Datadog by Ephasme... <3
> Press C-c to quit."
`;
